sphinx/tutorial/rose/furthertopics/failif-warnif.rst

```diff
@@ -0,0 +1,240 @@
+
+This states the relationship between these settings (a rearrangement
+of the `Tsiolkovsky rocket equation`_). The rocket must have a
+sufficient ratio of fuel to rocket mass, with a sufficiently fast
+exhaust velocity (``=9.8 * namelist:rocket=specific_impulse_s``)
+to get to the orbital speed ``env=ORBITAL_SPEED_MS``.
+
+Save the metadata file and then reload the config editor metadata.
+You now need to ask Rose to evaluate the ``fail-if`` condition, as
+it's an on-demand process.
+
+Either press the toolbar button :guilabel:`Check fail-if ...` or click
+the menu :menuselection:`Metadata --> Check fail-if, warn-if`.
+
+Hopefully, this should not flag any errors, as these are the Apollo
+mission parameters!
+
+Try adding a few more moonrocks. Add ``1000`` to the values
+of ``total_weight_kg`` and ``fuelless_weight_kg``.
+
+Re-run the check by clicking
+:menuselection:`Metadata --> Check fail-if, warn-if`. An error
+dialog will appear, and the ``total_weight_kg`` setting will have
+an error flag.
+
+However, neither of these are very informative, other than quoting
+the metadata.
+
+Change the ``fail-if`` line to:
+
+.. code-block:: rose
+
+   fail-if=this < namelist:rocket=fuelless_weight_kg * 2.7183**(env=ORBITAL_SPEED_MS / (9.8 * namelist:rocket=specific_impulse_s));  # Fuel mass ratio or specific impulse too low to achieve orbit.
+
+If you reload the metadata and run the check again, the error
+messages will include the helpful text.
+
+You can also check the ``fail-if`` metadata by running
+``rose macro --validate`` or ``rose macro -V`` in a terminal,
+inside the app directory. Try saving the configuration in a failed
+state, and then run the command.
+
+
+``warn-if``
+-----------
+
+The ``warn-if`` metadata setting is exactly the same as ``fail-if``,
+but gives out warnings instead of errors - from a user perspective,
+this means something like "Watch out" rather than "This is wrong".
+
+Let's try adding something for ``namelist:rocket=battery_levels``.
+
+Open the metadata file ``meta/rose-meta.conf`` in a text editor,
+and add this line to the ``[namelist:rocket=battery_levels]`` section:
+
+.. code-block:: rose
+
+   warn-if=namelist:rocket=battery_levels(1) < 75 or namelist:rocket=battery_levels(2) < 75;
+
+This uses a special syntax for referencing the individual array
+elements in ``battery_levels``.
+
+If the first array element value and/or the second array element
+value of ``battery_levels`` is less than 75% full, a warning will
+be produced when the check is run.
+
+We already know the shorthand syntax this, so rephrase the metadata to:
+
+.. code-block:: rose
+
+   warn-if=this(1) < 75 or this(2) < 75;
+
+Save the metadata file and then reload the config editor metadata.
+Click :menuselection:`Metadata --> Check fail-if, warn-if` - a
+warning dialog should appear, and there should be a warning flag
+for the ``battery_levels`` option.
```

Comment:
We get this as well as the 'error' for total_weight_kg from the previous section 'fail-if' - it is not an issue but it might be worth either stating something along the lines of 'as well as the error from the previous section' here or, at the start of the 'warn-if' section, asking readers to rewrite the app directory to get it back to its original (loaded) state or undo the few changes made in the 'fail-if' section.